### PR TITLE
Fix compiler crashing when on-fail used within anonymous definitions

### LINKED ISSUE #36067: NPE at BIROptimizer with on-fail within a resource function
**Description:**
```ballerina
import ballerina/websocket;
import ballerina/io;

listener websocket:Listener wsMgtListener = new websocket:Listener(9092);

map<websocket:Caller[]> riderRequestedLocationMap = {};

service /rider on wsMgtListener {
    resource isolated function get .(string name)
                    returns websocket:Service|websocket:Error {
        return service object websocket:Service {
            remote function onOpen(websocket:Caller caller) returns websocket:Error? {

                do {
                    check foo();
                } on fail error e {
                    io:println("error##", e);
                }

            }
            remote function onClose(websocket:Caller caller, int statusCode, string reason) {
                //_ = clientsMap.remove(caller.getConnectionId());
            }

            remote function onTextMessage(websocket:Caller caller, string buildingId) returns error? {
                io:println("rider call: ", buildingId);
                boolean isCallers = riderRequestedLocationMap.hasKey(buildingId);
                if (!isCallers) {
                    riderRequestedLocationMap[buildingId] = [caller];
                } else {
                    riderRequestedLocationMap.get(buildingId).push(caller);
                }
            }

        };
    }

}

function foo() returns error? {
    return error("Dummy error");
}
```

**Steps to reproduce:**

**Affected Versions:**
2201.0.3

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #36158: Compiler crashes with `on fail` in anonymous function expression
**Description:**
$title, when the error is used within the `on fail` clause.

**Steps to reproduce:**
```ballerina
import ballerina/log;

function fn() returns error? {
    stream<string[], error?> str = new;

    check str.forEach(function (string[] arr) {
        do {
            int _ = check int:fromString(arr[0]);
            int _ = check int:fromString(arr[1]);
        } on fail error e {
            // log:printError("Error"); // OK if `e` is not used
            log:printError("Error", e);
        }
    });
}
```

Workaround is to explicitly define the function.

```ballerina
import ballerina/log;

function fn() returns error? {
    stream<string[], error?> str = new;

    check str.forEach(argFn);
}

function argFn(string[] arr) {
    do {
        int _ = check int:fromString(arr[0]);
        int _ = check int:fromString(arr[1]);
    } on fail error e {
        log:printError("Error", e);
    }
}
```

**Affected Versions:**
2201.1.0-rc1

**Related Issues (optional):**
https://github.com/ballerina-platform/ballerina-lang/issues/36067

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #36067
Fixes #36158

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal [hunk 1]
```diff
@@ -357,6 +360,138 @@ function getError() returns error {
...
+                    return "Error thrown from on-fail: " + e2.message();
+                }
+                return "should not reach here";
+            }
+        };
+    }
+}
+
+function getService() returns object {} = @java:Method {
+    'class: "org/ballerinalang/nativeimpl/jvm/servicetests/ServiceValue",
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal:422]: @rdhananjaya is it okay to use the same `ServiceValue` in this new test case? I slightly remember you said there could be a conflict. So need to use a new one?

**Review by lasinicl — COMMENTED** [review]
Tested. Changes LGTM.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal (hunk 1 above)
Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal:422]: The same `ServiceValue` is being atm in different tests atm though

1. https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/test-src/services/service_decl.bal
2. https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/object/test_service_objects.bal
3. https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_service_class.bal)

Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal:422]: ServiceValue class is a static one so the test cases need to make sure that tests that use the ServiceValue do not run concurrently and make sure that ServiceValue is reset before and after the test.

Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal:422]: Each pkg in jbal-unit-tests runs consecutively (https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/testng.xml#L24). These tests are under a separate pkg(https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/testng.xml#L105). `ServiceValue` is reset at the beginning and end. Hence, should be fine then. Thanks

**Review by lochana-chathura — APPROVED** [review]

**Review by lasinicl — APPROVED** [review]

**Pull request merged**